src/core/eme/get_session.ts

```diff
@@ -17,110 +17,147 @@
 
 /**
  * Handle MediaEncryptedEvents sent by a HTMLMediaElement:
- * Either create a session, recuperate a previous session and returns it or load
- * a persistent session.
- * @param {Event} encryptedEvent
+ * Either create a MediaKeySession, recuperate a previous MediaKeySession or
+ * load a persistent session.
+ *
+ * Some previously created MediaKeySession can be closed in this process to
+ * respect the maximum limit of concurrent MediaKeySession, as defined by the
+ * `EME_MAX_SIMULTANEOUS_MEDIA_KEY_SESSIONS` config property.
+ *
+ * You can refer to the events emitted to know about the current situation.
+ * @param {Event} initializationDataInfo
  * @param {Object} handledInitData
  * @param {Object} mediaKeysInfos
  * @returns {Observable}
  */
 export default function getSession(
-  encryptedEvent : IEncryptedEvent,
+  initializationDataInfo : IInitializationDataInfo,
   mediaKeysInfos : IMediaKeysInfos
-) : Observable<IHandledEncryptedEvent> {
-  return observableDefer(() : Observable<IHandledEncryptedEvent> => {
-    const { type: initDataType, data: initData } = encryptedEvent;
+) : Observable<IGetSessionEvent> {
+  return observableDefer(() : Observable<IGetSessionEvent> => {
+    const { type: initDataType, data: initData } = initializationDataInfo;
 
-    // possible previous loaded session with the same initialization data
-    let previousLoadedSession : MediaKeySession|ICustomMediaKeySession|null = null;
-    const { sessionsStore } = mediaKeysInfos;
-    const entry = sessionsStore.get(initData, initDataType);
-    if (entry != null) {
-      previousLoadedSession = entry.session;
+    /**
+     * Store previously-loaded MediaKeySession with the same initialization data, if one.
+     */
+    let previousLoadedSession : MediaKeySession |
+                                ICustomMediaKeySession |
+                                null = null;
+
+    const { loadedSessionsStore } = mediaKeysInfos;
+    const entry = loadedSessionsStore.get(initData, initDataType);
+    if (entry !== null) {
+      previousLoadedSession = entry.mediaKeySession;
       if (isSessionUsable(previousLoadedSession)) {
         log.debug("EME: Reuse loaded session", previousLoadedSession.sessionId);
         return observableOf({ type: "loaded-open-session" as const,
                               value: { mediaKeySession: previousLoadedSession,
                                        sessionType: entry.sessionType,
                                        initData,
                                        initDataType } });
-      } else if (mediaKeysInfos.sessionStorage != null) {
-        mediaKeysInfos.sessionStorage.delete(new Uint8Array(initData), initDataType);
+      } else if (mediaKeysInfos.persistentSessionsStore != null) {
+        // If the session is not usable anymore, we can also remove it from the
+        // PersistentSessionsStore.
+        // TODO Are we sure this is always what we want?
```

Comment:
The question here is : is it possible that session asks by itself a new licence when it is used again ?

Reply:
In most cases, I think that this code is the sensible thing to do:  we remove a persistent MediaKeySession which has no key considered as "usable".

As we already realized, there are cases where a valid usable persistent MediaKeySession has no key marked as "usable" which we consider to be a browser bug for now. As you said I don't think that the session will start asking a new license by itself. So here, it makes sense that we're removing it to make place for a new MediaKeySession. This is the safe way of doing it.

Here though that TODO is not referencing that case only.
This comment was more about the way things are written here.

By considering this code only, we just now that:
  1. a MediaKeySession corresponding to that initialization data was already loaded on the current `MediaKeys` instance (we know that because `loadedSessionsStore.get(initData, initDataType);` returned something)
  2. that session has no key considered as "usable" (we know that because `isSessionUsable(previousLoadedSession)` returned false)

Based on that, we remove any stored persistent MediaKeySession with that initialization data.

I feel like we're missing an element here. By just reading that code in isolation, nothing proves that this MediaKeySession has anything to do with a persisted one with the same initialization data.

So, that's not a bug or anything for now, it's just not clear.
I could update the comment if you think I should do that.

Reply:
Maybe we could be more explicit about the fact that there may be no persistent session associated with this initData.

Reply:
Yes but more than that, it's "the persistent MediaKeySession we are removing might not be the same one"